text/0000-unsafe-fields.md

```diff
@@ -0,0 +1,137 @@
+- The elements at `old_len..new_len` must be initialized.
+
+This field is safe to read, but unsafe to mutate or initialize due to the invariants. These
+invariants cannot be expressed in the type system, so they must be enforced manually. Failure to do
+so may result in undefined behavior elsewhere in `Vec`.
+
+By introducing unsafe fields, Rust can improve the situation where a field that is otherwise safe is
+used as a safety invariant.
+
+# Guide-level explanation
+
+Fields may be declared `unsafe`. Unsafe fields may only be initialized or accessed mutably in an
+unsafe context. Reading the value of an unsafe field may occur in either safe or unsafe contexts. An
+unsafe field may be relied upon as a safety invariant in other unsafe code.
+
+Here is an example to illustrate usage:
+
+```rust
+struct Foo {
+    safe_field: u32,
+    /// Safety: Value must be an odd number.
+    unsafe unsafe_field: u32,
+}
+
+// Unsafe field initialization requires an `unsafe` block.
+// Safety: `unsafe_field` is odd.
+let mut foo = unsafe {
+    Foo {
+        safe_field: 0,
+        unsafe_field: 1,
+    }
+};
+
+// Safe field: no unsafe block.
+foo.safe_field = 1;
+
+// Unsafe field with mutation: unsafe block is required.
+// Safety: The value is odd.
+unsafe { foo.unsafe_field = 3; }
+
+// Unsafe field without mutation: no unsafe block.
+println!("{}", foo.unsafe_field);
+```
+
+For a full description of where a mutable access is considered to have occurred (and why), see
+[RFC 3323]. Keep in mind that due to reborrowing, a mutable access of an unsafe field is not
+necessarily explicit.
+
+[RFC 3323]: https://rust-lang.github.io/rfcs/3323-restrictions.html#where-does-a-mutation-occur
+
+```rust
+fn change_unsafe_field(foo: &mut Foo) {
+    // Safety: An odd number plus two remains an odd number.
+    unsafe { foo.unsafe_field += 2; }
+}
+```
+
+# Reference-level explanation
+
+## Syntax
+
+Using the syntax from [the reference for structs][struct syntax], the change needed to support
+unsafe fields is minimal.
+
+[struct syntax]: https://doc.rust-lang.org/stable/reference/items/structs.html#structs
+
+```diff
+StructField :
+   OuterAttribute*
+   Visibility?
++  unsafe?
+   IDENTIFIER : Type
+
+TupleField :
+   OuterAttribute*
+   Visibility?
++  unsafe?
+   Type
+```
+
+## Behavior
+
+An unsafe field may only be mutated or initialized in an unsafe context. Failure to do so is a compile error.
+
+## "Mutable use" in the compiler
+
+The concept of a "mutable use" [already exists][mutating use] within the compiler. This catches all
+situations that are relevant here, including `ptr::addr_of_mut!`, `&mut`, and direct assignment to a
+field, while excluding interior mutability. As such, formal semantics of what constitutes a "mutable
+use" are not stated here.
+
+[mutating use]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/visit/enum.PlaceContext.html#method.is_mutating_use
+
+# Drawbacks
```

Comment:
I would add something about the proliferation of `unsafe` use cases, since a lot of programs that currently `#![deny(unsafe)]` without exception will likely need to dial it back.

This isn't necessarily a bad thing, but it does seem like it starts uplifting `unsafe` to a more general-user usecase (as opposed to mostly users reasoning about hardware interaction).

Reply:
If a crate uses no unsafe, then it has no need for unsafe fields.

Reply:
The proposal is guided toward more broadly preventing logic errors rather than strictly what is now known as UB, no? One of the examples linked above was keeping flags in sync with fields for a time parser, which I don't think was tied to any existing `unsafe` calls.

Reply:
> One of the examples linked above was keeping flags in sync with fields for a time parser

The `flags` field in that example serves as a bitpacked equivalent of the discriminant for `Option`. Those flags are used to determine whether it's safe to read a `MaybeUninit` value. It's 100% tied to memory safety.

Reply:
I stand corrected on the flags case - but the RFC example is using these fields to ensure that a value is odd with no other `unsafe` interaction. If non-UB cases aren't a target of the RFC, could you update the example to show something where not upholding unsafe field invariants would lead to UB?

Reply:
That's fair. The example of an even number isn't the best, as it's contrived. There's an unstated assumption that it's being relied upon as a safety invariant somewhere else in the program. I can definitely improve that example.